# Karyotype alteration in Colorectal cancer metastasis <img align="right" width="150" height="100" src=/images/Barts.jpg>


The aim is to determine if the seeding of a primary colorectal cancer to a distant metastatic site results in specific karyotype adjustments that lead to adaptation to a new tissue-specific microenvironment.

## Description
Cancer cells carry different types of mutations associated with cell proliferation and cell longevity. 
Certain alterations only affect one or a few base pairs. Others, known as copy number alterations (CNAs), involve larger portions of the genome that can either be deleted or gained. 
Tumours in different patients carry variable amounts of these deletions or amplifications. Genome-wide CNA profiles, or karyotypes, are in many cases characteristic of the tissue of origin. 
It is known that tumour CNA burden is a pan-cancer prognostic factor associated with recurrence and that genome doubling (GD) is associated with metastasis. 
However, it is unknown if the seeding of a primary tumour to a distant metastatic site results in specific karyotype adjustments that facilitate adaptation to the new microenvironment.

## Dependencies
```python
import pandas as pd
import numpy as np
import os
import warnings
```
```R
library("openxlsx", "tidyverse", "survminer", "ComplexHeatmap", "RColorBrewer",
"ggbeeswarm", "MASS","grid", "gtable", "gsubfn", "proto", "circlize", "xlsx", 
"RColorBrewer", "ComplexHeatmap", "grid","data.table", "dplyr", "survminer", 
"ggpubr", "magrittr", "survival", "openxlsx", "reshape2", "reshape","copynumber",
"BiocGenerics", "parallel", "ggplot2", "stats", "graphics", "grDevices", "utils", 
"datasets", "methods", "base", "tidyr","tidyverse","spatstat")
```
## Versions
### R 4.0.5
### Python 3.8.10

## Analysis
### Data Collection
#### ReadHartwigFiles.py
This script is used to extract somatic copy number calls from metadata
generated by hartwig Medical foundation.

### Data Transformation
#### final.py
This script is used to bin Hartwig data into 1MB size.


#### tcgaBins.py
This script is used to bin TCGA data into 1MB size.


#### pandqarmscalculation.R
This script is used to annotate bins belonging to p and q arms.


#### pandqdf_script.R
This script is used to create a dataframe with p and q arms
annotation to each chromosome.

### Statistical tests

#### heatmap_df_script.R
This script is used to create the dataframe from Hartwig copy number calls
for complex heatmap.


#### heatmap_script.R
This script is used to render complex heatmap with annotations of biopsy
sites and category of aberrations.


#### plotFreq_script.R
This script is used to calculate frequency of loss and gains across the
genome. Rendering it into a plot for comparison between primary and 
metastatic sites.


#### quantifying_loss_gain.R
This script is used to quantify loss and gain percentage difference.
The students t-test is used to determine significant difference between
gains in primary and metastatic samples. Same mean test is used to 
quantify losses in primary and metastatic samples.


#### WGD_frequency_script.R
This script is used to annotate and calculate frequency of genome doubled
samples at each biopsy site in primaries and metastatic data. Rendered
using ggplot.


#### WGD_script.R
This script is used to remove genome doubling from primary as well as 
metastatic copy number data. Furthermore, it calculates loss and gains
across the genome and renders it into a plot for comparison between 
corrected primary and metastatic samples.


#### ViolinPlot.R
This script is used to compare genome doubled data vs non genome doubled
and render violin plot for same. It also compares fraction of aberration
between primary and metastatic sites where copy number calls are genome
doubled as well as copy number calls where genome doubling is corrected.
This script also covers statistical mean comparison.


#### fischer_script.R
This script is used to perform fisher exact test between lung and liver 
samples. Also, it covers FDR correction.


#### gene_annotation.R
This script is used to annotate genes to the genomic region which came
out as significant from the fisher test. These genes were then used for
pathway analysis using PANTHER web UI.


#### PCA_script.R
This script is performing Principal Component Analysis on combined copy
number calls for primary as well as metastatic data. ggplot2 is used here
to render PCA plots.


#### K-means_script.R
This script is performing K-means clustering on the combined copy number
calls from parimary and metastatic data. It also covers the codes used to
render a complex heatmap to visualise k-means cluster along with annotation
which reveals whether the samples are whole genome doubled or not.


#### MSI_status_script.R
This script is calculating percentage of MSI and MSS samples in the five 
clusters obtained in K-means clustering.





